### PR TITLE
makefiles: tools/renode: add support for Renode

### PR DESCRIPTION
**Pull request by mgielda**
Thanks to @basilfx's contribution, this is an integration with the [Renode framework](https://github.com/renode/renode), which adds `make emulate` as an additional target. 

Current execution is as follows: `make emulate` -> generate intermediate script (to include environment variables like `$ELFFILE`) -> start Renode using intermediate script.

Improving on the original integration, Renode can be installed from [packages for different OSs/distributions](https://github.com/renode/renode/releases/latest).

As for now, the `hello-world` example has been tested to work, while e.g. `gnrc_networking` is known to fail on an assert in the radio driver - this will be fixed soon.

### CODE REVIEW

**Comment by miri64** [comment]
Can you explain the decision for me to go for renode and against emul8?

**Comment by basilfx** [comment]
Renode is the successor of Emul8.

**Comment by miri64** [comment]
Would have been clearer if that was mentioned in the PR's description ;-)

**Comment by basilfx** [comment]
I had some inside information from @mgielda, but I agree :-)

Anyway, I've been able to run this example AND create an emulation from scratch for the STK3600:

schermafbeelding 2017-11-08 om 23 46 10

It's amazing what you can do when you have full control over the CPU. You can even attach GDB to your emulator, and step through the code.

**Comment by basilfx** [comment]
Since I've created the initial PR, I of course cannot ACK :-)

Supported boards is limited to the devices currently supported by Renode, but [that list](https://github.com/renode/renode/tree/master/platforms/cpus) could grow. I have plans for adding EFM32 support.

Any thoughts on the RIOT side of this PR (the `Makefile`, adding board definitions to `board/dist` folder) (cc @kaspar030)?

**Comment by basilfx** [comment]
Ping @OlegHahm @miri64.

**Comment by miri64** [comment]
I was able to run `hello-world` and `ipc_pingpong` with this (renode 1.1 from DEB package on Ubuntu 16.04) and they run as expected, but I still get the same broken newlines problem in the terminal emulator I had with emul8... Maybe I have configured something wrong, since @basilfx did not have that problem.

renode output

**Comment by basilfx** [comment]
Hmm, maybe @mgielda knows if it is platform dependent?

**Comment by mgielda** [comment]
Sorry, going forward we're making the CRLF sanitizing on by default, but in your installation it's still off (which is the old default) - please right-click on the console and set it to true ("Append '\r' to line ending"); overall we're making a new release today, I'll let you know once it's online.

**Comment by miri64** [comment]
Ok, that worked for me! :-).

File: dist/tools/renode/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+### From source
+Follow the installation instructions on Renode's [GitHub](https://github.com/renode/renode#installation) page.
+
+After compilation is successful, ensure that `renode` is available on your `PATH`. One way to do so, is via symlink: `sudo ln -s path/to/renode/repository/run.sh /usr/local/bin/renode`.
+
+### Testing
+After installation, verify if Renode is working using `renode --help`. You should be presented with a help screen.
+
+## Usage
+From within RIOT-OS, you can use `make emulate` to start emulation. It expects a board definition file in `boards/<BOARD>/dist/board.resc`.
```
Comment (miri64) [inline dist/tools/renode/README.md]: Is there any doc on how to write such a `board.resc` file? If yes, it might be a good idea to link it here.

Comment (basilfx) [reply dist/tools/renode/README.md]: According to https://github.com/renode/renode/, documentation is work in progress :-)

However, some work in progress documentation can be found here: https://renode.readthedocs.io/en/latest/.

Comment (miri64) [reply dist/tools/renode/README.md]: Even a link to a WIP documentation would be appreciated ;-).

**Comment by basilfx** [comment]
@mgielda Would you mind to add a link to the documentation? When this is merged, I am willing to add initial `board.resc` files for the CPUs currently supported.

**Comment by kaspar030** [comment]
How about using features for this? See https://github.com/kaspar030/RIOT/tree/pr_pr7959_use_feature_for_renode.

This way, we can do e.g., ```make info-boards-supported FEATURES_REQUIRED+=renode``` in order to find boards supporting renode.

**Comment by miri64** [comment]
@kaspar030 aren't features more about compiled-in features? I think in the case of `renode` (or any other emulator for that matter) the existence of the `emulate` make-target should suffice to hint about the emulation feature, IMHO.

**Comment by kaspar030** [comment]
> @kaspar030 aren't features more about compiled-in features? 

No, features are about "what features does the board support". One it is compiled in, it is either just there (e.g., cpp), or a module (e.g., MODULE_PERIPH_GPIO).

We can argue a lot about what "feature" means, and whether we should use it for MCU hardware features, only for configured MCU hardware features, for software features (cpp, ssp), this or that.
In the end, it provides a convenient (and established) way of specifying that a board supports, with all the nesting down to architecture level. And better, a way to use that information on the highest make level, before most of the board specific stuff has been parsed, thus for global make targets.

> I think in the case of renode (or any other emulator for that matter) the existence of the emulate make-target should suffice to hint about the emulation feature, IMHO.

The emulate-target (currently) always exists. There might be other emulators.
Also, based on the target, how would you do the equivalent of ```make info-boards-supported FEATURES_REQUIRED+=renode```?

**Comment by miri64** [comment]
Okay, I didn't know the semantics of FEATURES extended that much.

**Comment by mgielda** [comment]
Thanks guys. The release was published yesterday. I will add the link to the docs in a second, we're also adding some stuff as we speak. @kaspar030's proposition with features is quite cool but of course depends on your strategy for this.

@basilfx as for the EFM platforms, we'll be in contact how to do this properly for the new release.

**Comment by mgielda** [comment]
OK, added the link.

**Comment by miri64** [comment]
> but of course depends on your strategy for this.

I think my knowledge about FEATURE is just pretty outdated, so no strategy to discuss ;-) Go ahead as @kaspar030 proposed.

**Comment by mgielda** [comment]
Let's add the FEATURES thing as a follow up - I'd be happy if someone from RIOT (preferably @kaspar030) did that as a separate PR.

**Comment by miri64** [comment]
@kaspar030 are you okay with doing it as a follow-up? If yes, hit merge please.

**Comment by miri64** [comment]
Oops, doesn't need squashing. Did not realize at first that the commits to the README were of different authorship => can be merged as is (if @kaspar030 is okay with it of course)

**Comment by basilfx** [comment]
Ping @kaspar030

**Comment by miri64** [comment]
Since @kaspar030 is sick right now, I'm making a judgement call and merge this. If @kaspar030 is not willing to take on the follow-up, I probably can squeeze it in.

**Comment by miri64** [comment]
Since it only came down to one line, I did it right now: https://github.com/RIOT-OS/RIOT/pull/8137 ;-)

**Comment by mgielda** [comment]
Fantastic! Thanks :)

**Pull request merged**